crates/oxc_ast/src/serialize.rs

```diff
@@ -1021,6 +1021,57 @@ impl ESTree for TSModuleDeclarationGlobal<'_, '_> {
     }
 }
 
+/// Serializer for `key` and `constraint` field of `TSMappedType`.
+#[ast_meta]
+#[estree(
+    ts_type = "TSTypeParameter['name']",
+    raw_deser = "
+        const typeParameter = DESER[Box<TSTypeParameter>](POS_OFFSET.type_parameter);
+        typeParameter.name
+    "
+)]
+pub struct TSMappedTypeKey<'a, 'b>(pub &'b TSMappedType<'a>);
+
+impl ESTree for TSMappedTypeKey<'_, '_> {
+    fn serialize<S: Serializer>(&self, serializer: S) {
+        self.0.type_parameter.name.serialize(serializer);
+    }
+}
+// NOTE: Variable `typeParameter` in `raw_deser` is shared between `key` and `constraint` serializers.
+// They will be concatenated in the generated code.
+#[ast_meta]
+#[estree(ts_type = "TSTypeParameter['constraint']", raw_deser = "typeParameter.constraint")]
+pub struct TSMappedTypeConstraint<'a, 'b>(pub &'b TSMappedType<'a>);
+
+impl ESTree for TSMappedTypeConstraint<'_, '_> {
+    fn serialize<S: Serializer>(&self, serializer: S) {
+        self.0.type_parameter.constraint.serialize(serializer);
+    }
+}
+
+#[ast_meta]
+#[estree(
+    ts_type = "true | '+' | '-' | null",
+    raw_deser = "
+        const operator = uint8[pos];
+        [true, '+', '-', null][operator]
```

Comment:
And I don't even think this is the right way to do it. Sorry, could you please give me some guidance? 🙇🏻

Reply:
Applied in 089df24b877a684db25abd107b283fc18b3c6e62.